testsupport/assertions/wait.go

```diff
@@ -0,0 +1,194 @@
+package assertions
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/codeready-toolchain/toolchain-e2e/testsupport/wait"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func WaitFor[T client.Object](cl client.Client) *Finder[T] {
+	return &Finder[T]{
+		cl:      cl,
+		timeout: wait.DefaultTimeout,
+		tick:    wait.DefaultRetryInterval,
+	}
+}
+
+type Finder[T client.Object] struct {
+	cl      client.Client
+	timeout time.Duration
+	tick    time.Duration
+}
+
+type FinderInNamespace[T client.Object] struct {
+	Finder[T]
+	namespace string
+}
+
+type FinderByObjectKey[T client.Object] struct {
+	Finder[T]
+	key client.ObjectKey
+}
+
+type failureTrackingT struct {
+	*assert.CollectT
+	failed bool
+}
+
+func (f *Finder[T]) WithTimeout(timeout time.Duration) *Finder[T] {
+	f.timeout = timeout
+	return f
+}
+
+func (f *Finder[T]) WithRetryInterval(interval time.Duration) *Finder[T] {
+	f.tick = interval
+	return f
+}
+
+func (f *Finder[T]) ByObjectKey(namespace, name string) *FinderByObjectKey[T] {
+	return &FinderByObjectKey[T]{
+		Finder: *f,
+		key:    client.ObjectKey{Name: name, Namespace: namespace},
+	}
+}
+
+func (f *Finder[T]) InNamespace(ns string) *FinderInNamespace[T] {
+	return &FinderInNamespace[T]{
+		Finder:    *f,
+		namespace: ns,
+	}
+}
+
+func (f *FinderInNamespace[T]) WithName(name string) *FinderByObjectKey[T] {
+	return &FinderByObjectKey[T]{
+		Finder: f.Finder,
+		key:    client.ObjectKey{Name: name, Namespace: f.namespace},
+	}
+}
+
+func (t *failureTrackingT) Errorf(format string, args ...interface{}) {
+	t.failed = true
+	t.CollectT.Errorf(format, args...)
+}
```

Comment:
I don't see that you would implement `FailNow` anywhere. What if someone uses `require.*` in the predicates?

Reply:
`failureTrackingT` embeds `assert.CollectT` so it implements `FailNow` by the virtue of embedding.

Reply:
which means that it will kill only the "inner" execution inside of the `EventuallyWithT` and starts a new one, and not the whole test, right?